### ISSUE TITLE
Bug : LuceneIndex loses values (Date, Instant, ...) after mutation 

### ISSUE THREAD

**Issue by Lakedaemon** [JanusGraph/janusgraph#764]
I have been looking at the LuceneIndex code lately and I thought that it should be buggy regarding values that aren't stored when the graph gets mutated (they are not reindexed and so they are lost)

I exposed such a bug in the following test (sorry, I'm a kotlin developer, it is coded in kotlin and spek, but it should be easy to convert the test to java (I used gremlin))

I add a single vertex to a janusgraph with 2 values : a string and a date.
Those values are indexed
When I mutate the graph by changing the string value,
the index can't find the date anymore, it is lost.

(silent bug leading to performance/result issues)

My output in intellijidea : 
```
Testing started at 10:42 ...
173 [main] INFO org.janusgraph.graphdb.configuration.GraphDatabaseConfiguration - Set default timestamp provider MICRO
179 [main] INFO org.janusgraph.graphdb.configuration.GraphDatabaseConfiguration - Generated unique-instance-id=7f0001019750-mephistolus1
191 [main] INFO org.janusgraph.diskstorage.Backend - Configuring index [search]
453 [main] INFO org.janusgraph.diskstorage.Backend - Initiated backend operations thread pool of size 16
526 [main] INFO org.janusgraph.diskstorage.log.kcvs.KCVSLog - Loaded unidentified ReadMarker start time 2017-11-28T09:42:20.511Z into org.janusgraph.diskstorage.log.kcvs.KCVSLog$MessagePuller@616fe72b
530 [main] INFO org.janusgraph.diskstorage.Backend - Configuring index [search]
531 [main] INFO org.janusgraph.diskstorage.Backend - Initiated backend operations thread pool of size 16
608 [main] INFO org.janusgraph.graphdb.configuration.GraphDatabaseConfiguration - Set default timestamp provider MICRO
608 [main] INFO org.janusgraph.graphdb.configuration.GraphDatabaseConfiguration - Generated unique-instance-id=7f0001019750-mephistolus2
609 [main] INFO org.janusgraph.diskstorage.Backend - Configuring index [search]
614 [main] INFO org.janusgraph.diskstorage.Backend - Initiated backend operations thread pool of size 16
617 [main] INFO org.janusgraph.diskstorage.log.kcvs.KCVSLog - Loaded unidentified ReadMarker start time 2017-11-28T09:42:20.613Z into org.janusgraph.diskstorage.log.kcvs.KCVSLog$MessagePuller@1b2c4efb
1
2798 [main] WARN org.janusgraph.graphdb.transaction.StandardJanusGraphTx - Query requires iterating over all vertices [()]. For better performance, use indexes
dateProperty:[Thu Jan 08 18:11:06 CET 1970]
stringProperty:[a]
hit:true  // <--- before mutation, the date can be found  in the index
1'
_____elementid:37c 1 -1
stringProperty:a 1 -1
2832 [main] WARN org.janusgraph.graphdb.transaction.StandardJanusGraphTx - Query requires iterating over all vertices [()]. For better performance, use indexes
2837 [main] WARN org.janusgraph.graphdb.transaction.StandardJanusGraphTx - Query requires iterating over all vertices [()]. For better performance, use indexes
2
hit:false  // <--- after mutation, the date cannot be found in the index
_____elementid:37c 1 -1
stringProperty:b 1 -1

Process finished with exit code 0
```

```kotlin
package org.lakedaemon.graph

import org.apache.lucene.document.LongPoint
import org.apache.tinkerpop.gremlin.structure.T
import org.apache.tinkerpop.gremlin.structure.Vertex
import org.janusgraph.core.Cardinality
import org.janusgraph.core.JanusGraphFactory
import org.janusgraph.core.schema.Mapping
import org.jetbrains.spek.api.Spek
import org.jetbrains.spek.api.dsl.describe
import org.jetbrains.spek.api.dsl.it
import org.jetbrains.spek.api.dsl.on
import java.io.File
import org.apache.lucene.index.DirectoryReader
import org.apache.lucene.index.MultiFields
import org.apache.lucene.search.IndexSearcher
import org.apache.lucene.store.NIOFSDirectory
import org.apache.tinkerpop.gremlin.structure.Graph
import java.util.*

val indexRootPath = "/home/lakedaemon/janusgraphIndex"
val indexPath = "/home/lakedaemon/janusgraphIndex/custom"
object JanusgraphBugSpek : Spek({

    describe("mutate JanusGraph") {
        on("lucene index") {
            it("should keep data fields") {
                val builder = JanusGraphFactory.build().apply {
                    set("storage.backend", "inmemory")
                    set("index.search.backend", "lucene")
                    set("index.search.directory", indexRootPath)
                    set("schema.default", "none")
                }

                JanusGraphFactory.drop(builder.open())
                val graph = builder.open()
                graph.let {
                    it.tx().rollback()
                    /** labels/properties & types */
                    val mgmt = it.openManagement()
                    mgmt.makeVertexLabel(VERTEX).make()

                    /** i18n */
                    val stringProperty = mgmt.makePropertyKey(stringPropertyKey).cardinality(Cardinality.SINGLE).dataType(String::class.java).make()
                    val dateProperty = mgmt.makePropertyKey(datePropertyKey).cardinality(Cardinality.SINGLE).dataType(Date::class.java).make()

                    mgmt.buildIndex("custom", Vertex::class.java)
                            .addKey(stringProperty, Mapping.STRING.asParameter())
                            .addKey(dateProperty)
                            .buildMixedIndex("search")

                    mgmt.commit()
                }

                graph.tx().open()
                val vertex = graph.addVertex(T.label, VERTEX)
                vertex.property(stringPropertyKey, "a")
                vertex.property(datePropertyKey, date)
                graph.tx().commit()
                //graph.traversal().V().map({System.out.println(it.get().property<Date>(datePropertyKey).toString())})

                System.out.println("1")
                graph.exposeVertex()
System.out.println("hit:"+hasHitForDate())
                //graph.traversal().V().has(datePropertyKey, date).map({System.out.println(it.get().property<Date>(datePropertyKey).toString())})
                System.out.println("1'")
                exposeFields()

                graph.traversal().V().properties<String>(stringPropertyKey).drop().iterate()
                graph.traversal().V().property(stringPropertyKey, "b").iterate()
                graph.tx().commit()
                System.out.println("2")
                System.out.println("hit:"+hasHitForDate())
                //graph.traversal().V().has(datePropertyKey, Date(666666666L)).map({System.out.println(it.get().property<Date>(datePropertyKey).toString())})
                graph.traversal().V().map({System.out.println(it.get().property<Date>(datePropertyKey).toString())})

                exposeFields()
                graph.close()
            }
        }
    }


})
val VERTEX = "VERT"
val stringPropertyKey = "stringProperty"
val datePropertyKey = "dateProperty"
val dateValue = 666666666L
val date = Date(dateValue)

fun Graph.exposeVertex() {
    traversal().V().map({
        val v = it.get()
        for (key in v.keys()) java.lang.System.out.println("$key:${v.properties<Any>(key).asSequence().map({it.value().toString()}).toList()}")
    }).next()
}

fun exposeFields() {
    val directory = DirectoryReader.open(NIOFSDirectory.open(File(indexPath).toPath()))
    val fields = MultiFields.getFields(directory)
    for (field in fields) {
        val terms = fields.terms(field)
        val termEnum = terms.iterator()
        while (true) {
            val byteRef = termEnum.next() ?: break
            System.out.println("$field:${byteRef.utf8ToString()} ${termEnum.docFreq()} ${termEnum.totalTermFreq()}")
        }
    }
    directory.close()
}


fun hasHitForDate() :Boolean {
    val directory = DirectoryReader.open(NIOFSDirectory.open(File(indexPath).toPath()))
    val searcher = IndexSearcher(directory)
    val result = searcher.search(LongPoint.newExactQuery(datePropertyKey, dateValue),1)
    directory.close()
    return result.totalHits != 0L
}
```

**Comment by robertdale** [comment]
The above example demonstrates a drop/create mutation but updates `property(key,val)` are also impacted. It appears that something short-circuits the update.  That is, if the first key is mutated, then the following keys in the index are **not** also updated. However, if the last (or second, in this case) key of the index is updated, then all keys appear to be retained and queries work as expected.

Here's a simplified test case that can run in the console:
```
graph = JanusGraphFactory.build().
set("storage.backend","inmemory").
set("index.search.backend","lucene").
set("index.search.directory","indexRootPath").
set("schema.default","none").
open()
g = graph.traversal()
mgmt = graph.openManagement()
mgmt.makeVertexLabel('vert').make()
stringProperty = mgmt.makePropertyKey('stringPropertyKey').cardinality(Cardinality.SINGLE).dataType(String.class).make()
dateProperty = mgmt.makePropertyKey('datePropertyKey').cardinality(Cardinality.SINGLE).dataType(Long.class).make()
mgmt.buildIndex("custom", Vertex.class).addKey(stringProperty, Mapping.STRING.asParameter()).addKey(dateProperty).buildMixedIndex("search")
mgmt.commit()
g.addV('vert').property('stringPropertyKey','a').property('datePropertyKey', 6L)
g.tx().commit()
g.V().has('stringPropertyKey','a').has('datePropertyKey', 6L)
g.V().has('stringPropertyKey','a').has('datePropertyKey', 6L).property('stringPropertyKey','b')
g.tx().commit()
g.V().has('stringPropertyKey','b').has('datePropertyKey', 6L)
g.V().has('stringPropertyKey','b')
g.V().property('datePropertyKey', 7L)
g.tx().commit()
g.V().has('stringPropertyKey','b').has('datePropertyKey', 7L)
g.V().has('stringPropertyKey','b')
g.V().has('datePropertyKey', 7L)
```

**Comment by Lakedaemon** [comment]
Thanks for the confirmation.
There are more details (and much rambling) about this in issue #670 
As I need Set and LIST cardinality for lucene index as well as custom Analyzer support for Lucene indexes, I might fix this bug too and send a pr in a few days/weeks

**Comment by Lakedaemon** [comment]
Ok, I have written a fix for this issue and implemented Cardinal.LIST and Cardinal.SET for lucene.
There has been quite some refactoring going on.
I'll take some time to clean my code, check that I didn't mess GeoShapes with PointVectorStrategy, and probably try to make a map construction lazy for performance purposes, and then I'll do a pull request

I have been a kotlin developer for the past 3/4 years and my java skills are a bit rusty. 
So, my java code might feel a bit weird. Sorry about that (it is now so painfull to write java code...)

**Comment by Lakedaemon** [comment]
By the way, is it possible that there might be a possible problem with my fix for Cardinality.LIST :
 
When you mutate list(a, b, a, c)  to list(a, b, c).  what kind of mutation is sent ? 
(the index has no way to know that it is the second a that was removed and not the first a)
I'm not sure what I should do when receiving a deletion with value "a" 
(IMO mutations are badly designed for re-indexing)

**Comment by Lakedaemon** [comment]
ok, I have started working on a pr for that

**Comment by Lakedaemon** [comment]
my pr is nearly ready. It passes all tests except this one (testIndexUpdatesWithoutRedindexing, in LuceneBerkeleyTest)
 if (indexFeatures.supportsCardinality(Cardinality.LIST)) {
            evaluateQuery(tx.query().has("phone", Cmp.EQUAL, "1234"),
                    ElementCategory.VERTEX, 2, new boolean[]{true, true}, "theIndex");

I would like to make it pass but I just can't understand what mechanism it uses.
It looks like it doesn't even uses the lucene index and queries...

Can someone tell me what it does ? and how it evaluates the index ? please ?

**Comment by Lakedaemon** [comment]
I resolved my issue (previosly, I wasn't working on the latest master commit).
Now, I'm struggling with GEO shapes :/

**Comment by porunov** [comment]
Fixed by #1645
